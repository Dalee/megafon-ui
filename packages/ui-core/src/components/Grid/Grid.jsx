import * as React from 'react';
import * as PropTypes from 'prop-types';
import './Grid.less';
import cnCreate from 'utils/cn';

const cn = cnCreate('mfui-grid');
class Grid extends React.Component {
    static propTypes = {
        as: PropTypes.oneOf(['ul', 'ol']),
        hAlign: PropTypes.oneOf(['left', 'center', 'between', 'around']),
        vAlign: PropTypes.oneOf(['top', 'bottom', 'center', 'baseline']),
        guttersLeft: PropTypes.oneOf(['large', 'medium']),
        guttersBottom: PropTypes.oneOf(['large', 'medium']),
        className: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
        ]).isRequired,
    };

    static defaultProps = {
        multiRow: true,
    };

    render() {
        const { guttersLeft, guttersBottom, multiRow, hAlign, vAlign } = this.props;

        return (
            <div className={cn()}>
                <div
                    className={cn('container', {
                        'multi-row': multiRow,
                        'h-align': hAlign,
                        'v-align': vAlign,
                        'gutters-left': guttersLeft,
                        'gutters-bottom': guttersBottom,
                    })}>
                        {React.Children.map(children, child =>
                            React.cloneElement(child, {
                                className: cn(
                                    'column',
                                    {
                                        'gutter-left': guttersLeft,
                                        'gutter-bottom': guttersBottom,
                                    },
                                    child.props.className
                                ),
                            })
                        )}
                </div>
            </div>
        );
    }
}

export default Grid;
