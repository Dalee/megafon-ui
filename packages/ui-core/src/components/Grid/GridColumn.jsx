import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cn';

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component {
    static propTypes = {
        wide: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        desktop: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        tablet: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        mobile: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.boolean,
        all: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: 12,
        flex: false,
        grow: false,
    };

    render() {
        const { all, wide, desktop, tablet, className, flex, grow, align, children, mobile } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
