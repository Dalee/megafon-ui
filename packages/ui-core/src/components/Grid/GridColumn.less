@import '~styles/base.less';

.grid-column {
    &_grow {
        flex-grow: 1;
    }
    &_align_right {
        margin-left: auto;
    }
    &_align_left {
        margin-right: auto;
    }
    &_align_center {
        margin-right: auto;
        margin-left: auto;
    }

    &_flex {
        display: flex;
        flex-shrink: 0;
        flex-grow: 0;
    }

    .generateColumns(all);

    @media @desktopMU {
        .generateColumns(wide);
    }
    @media @desktopS {
        .generateColumns(desktop);
    }
    @media @mobileB {
        .generateColumns(tablet);
    }
    @media @mobileSM {
        .generateColumns(mobile);
    }
}

.generateColumns(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_@{resolution}_@{i} {
        width: (100% / @columnCount) * @i;
    }
    .generateColumns(@resolution, @columnCount, (@i + 1));
}
