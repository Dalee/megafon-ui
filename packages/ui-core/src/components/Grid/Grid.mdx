---
name: Grid
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Grid from './Grid';
import GridColumn from './GridColumn';

# Grid

## Properties

<PropsTable of={Grid} />

# GridColumn

## Properties

<PropsTable of={GridColumn} />

## Basic usage

<Playground>
    <Grid>
        <GridColumn>Item1</GridColumn>
        <GridColumn>Item2</GridColumn>
        <GridColumn>Item3</GridColumn>
        <GridColumn>Item4</GridColumn>
    </Grid>
</Playground>
