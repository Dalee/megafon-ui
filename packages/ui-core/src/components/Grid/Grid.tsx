import * as React from 'react';
import * as PropTypes from 'prop-types';
import './Grid.less';
import cnCreate from 'utils/cnCreate';
import { IGridColumn } from './GridColumn';

interface IGridProps {
    /** Alignment of all columns by horizontal axis */
    hAlign?: 'left' | 'right' | 'center' | 'between' | 'around';
    /** Alignment of all columns by vertical axis */
    vAlign?: 'top' | 'bottom' | 'center' | 'baseline';
    /** Margin area on the left side of column */
    guttersLeft?: 'large' | 'medium';
    /** Margin area on the bottom side of column */
    guttersBottom?: 'large' | 'medium';
    /** Transfering of columns onto multiple lines */
    multiRow?: boolean;
    children: Array<React.ReactElement<IGridColumn>> | React.ReactElement<IGridColumn>;
}

const cn = cnCreate('mfui-beta-grid');
class Grid extends React.Component<IGridProps> {
    static propTypes = {
        hAlign: PropTypes.oneOf(['left', 'right', 'center', 'between', 'around']),
        vAlign: PropTypes.oneOf(['top', 'bottom', 'center', 'baseline']),
        guttersLeft: PropTypes.oneOf(['large', 'medium']),
        guttersBottom: PropTypes.oneOf(['large', 'medium']),
        multiRow: PropTypes.bool,
        children: PropTypes.node,
    };

    static defaultProps = {
        multiRow: true,
    };

    render() {
        const { children, guttersLeft, guttersBottom, multiRow, hAlign, vAlign } = this.props;

        return (
            <div className={cn('')}>
                <div
                    className={cn('container', {
                        'multi-row': multiRow,
                        'h-align': hAlign,
                        'v-align': vAlign,
                        'gutters-left': guttersLeft,
                        'gutters-bottom': guttersBottom,
                    })}>
                    {React.Children.map(children, (child: React.ReactElement<IGridColumn>) =>
                        React.cloneElement(child, {
                            className: cn(
                                'column',
                                {
                                    'gutter-left': guttersLeft,
                                    'gutter-bottom': guttersBottom,
                                },
                                child.props.className
                            ),
                        })
                    )}
                </div>
            </div>
        );
    }
}

export default Grid;
