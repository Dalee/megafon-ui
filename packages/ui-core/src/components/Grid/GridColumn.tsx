import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cn';

export interface IGridColumn {
    /** Size of columns on wide screens */
    wide?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on desktop */
    desktop?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on tablet */
    tablet?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on mobile */
    mobile?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Size of columns on all screens */
    all?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';
    /** Align columns */
    align?: 'right' | 'left' | 'center';
    /** Column flex grow */
    grow?: boolean;
    /** Column as flex container */
    flex?: boolean;
    /** Custom class name */
    className?: string;
    children: JSX.Element[] | Element[] | JSX.Element | string | Element;
}

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component<IGridColumn, {}> {
    static propTypes = {
        wide: PropTypes.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        desktop: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        tablet: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        mobile: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        all: PropTypes.oneOfType(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.bool,
        className: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: 12,
        flex: false,
        grow: false,
    };

    render() {
        const { all, wide, desktop, tablet, className, flex, grow, align, children, mobile } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
