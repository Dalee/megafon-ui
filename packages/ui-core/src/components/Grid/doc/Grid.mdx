---
name: Grid
menu: Компоненты
route: components/grid
---

import { Props } from 'docz';
import Grid from '../Grid';
import GridColumn from '../GridColumn';
import Example from './Grid.example.mdx';

# Grid

<DoczTabs>
    <Example />
    <div>
        <Props of={Grid} />
        <h2>GridColumn</h2>
        <Props of={GridColumn} />
    </div>
</DoczTabs>
