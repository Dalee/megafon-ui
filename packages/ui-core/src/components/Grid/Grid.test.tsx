import * as React from 'react';
import { shallow } from 'enzyme';
import Grid from './Grid';
import GridColumn from './GridColumn';

describe('<Grid />', () => {
    it('it renders Grid', () => {
        const wrapper = shallow(
            <Grid
                hAlign="right"
                vAlign="top"
                guttersLeft="large"
                guttersBottom="medium"
            >
                <GridColumn>child1</GridColumn>
                <GridColumn>child2</GridColumn>
            </Grid>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
