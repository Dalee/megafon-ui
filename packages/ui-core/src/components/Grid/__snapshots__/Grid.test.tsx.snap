// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> it renders Grid with props 1`] = `
<div
  className="mfui-grid"
>
  <div
    className="mfui-grid__container mfui-grid__container_h-align_right mfui-grid__container_v-align_top mfui-grid__container_gutters-left_large mfui-grid__container_gutters-bottom_medium"
  >
    <GridColumn
      all="12"
      className="mfui-grid__column mfui-grid__column_gutter-left_large mfui-grid__column_gutter-bottom_medium"
      flex={false}
      grow={false}
      key=".0"
    >
      child1
    </GridColumn>
    <GridColumn
      all="12"
      className="mfui-grid__column mfui-grid__column_gutter-left_large mfui-grid__column_gutter-bottom_medium"
      flex={false}
      grow={false}
      key=".1"
    >
      child2
    </GridColumn>
  </div>
</div>
`;

exports[`<Grid /> it renders Grid without props 1`] = `
<div
  className="mfui-grid"
>
  <div
    className="mfui-grid__container mfui-grid__container_multi-row"
  >
    <GridColumn
      all="12"
      className="mfui-grid__column"
      flex={false}
      grow={false}
      key=".0"
    >
      child1
    </GridColumn>
    <GridColumn
      all="12"
      className="mfui-grid__column"
      flex={false}
      grow={false}
      key=".1"
    >
      child2
    </GridColumn>
  </div>
</div>
`;
